[FEATURE] Style management re-work and upgrade

nirvn opened the pull request:
This PR adds a favorite feature to the style manager and improves user experience across the board. The big changes are:
* A new favorite grouping system was added, which the symbols list widget defaults to
* Users can now directly tag, as well as add to favorites, symbols *while saving those* to the style database
* The selected tag / smartgroup in the symbols list widget now persists when switching layers (and across sessions; the symbols list widget will update the tag / smartgroup combo box when users add / rename / remove categories
* To streamline style management, groups have been removed and fully replaced by tags
* Tags have been integrated into the import/export user interface

_(I've had difficulty ordering the changes in order of importance 😄)_

Until now, while QGIS had a solid style management feature, it's interface wasn't helping users make the most out of it. The absence of settings beyond a name when saving symbols has been a long-standing gap in that regard. This PR improves the situation by adding a style/symbol save dialog, through which users can immediately manage symbols as he/she in the process of adding symbols to his/her database:
![untitled2](https://cloud.githubusercontent.com/assets/1728657/20243908/f02efb7c-a99f-11e6-94ce-6fe9f867ee39.png)

The other big change in the PR is the streamlining of the way the symbols can be re-grouped together by removing groups and upgrading various parts of QGIS to rely on a better-exposed tags. 

The symbols list widget now shows a list of tags and smartgroups:
![untitled](https://cloud.githubusercontent.com/assets/1728657/20243928/71d5d3ee-a9a0-11e6-8fa3-2c1ef6eda6f6.png)

The style manager has been updated to rely in tags as its primary mean of re-grouping symbols, integrating tags with its sidebar, its symbol right-click menu, etc.:
![untitled3](https://cloud.githubusercontent.com/assets/1728657/20243939/c3c68b94-a9a0-11e6-90da-ebced1884693.png)

Lastly, the import/export dialogs are now relying on tags:
![untitled4](https://cloud.githubusercontent.com/assets/1728657/20251069/7709156c-aa48-11e6-9232-9edc7184ce85.png)

![untitled5](https://cloud.githubusercontent.com/assets/1728657/20251072/7b5e28d2-aa48-11e6-8aab-ccffef8aa0f5.png)


Also, brief extra note on the favorites grouping feature: this will allow QGIS to ship a larger number of symbols by default while not overwhelming new users who will be served - by default - the favorite symbols through the symbols list widget. The goal is to answer a recurring request I've heard over the years, namely that QGIS doesn't ship with enough saved symbols.

*Important:* you will need to replace the symbology-nb.db with the updated one in this PR as the structure of a few tables were modified.

nirvn committed:
[FEATURE] Style management re-work and upgrade
- A new favorite grouping system was added, which the symbols list
widget defaults to
- The selected tag / smartgroup in the symbols list widget now
persists when switching layers (and across sessions)
- The symbols list widget will update the tag / smartgroup combo
box when users add / rename / remove categories
- Users can now directly tag, as well as add to favorites, symbols
while saving those to the style database
- To streamline style management, groups have been removed and
fully replaced by tags
- Tags have been integrated into the import/export user interface

nirvn commented:
@anitagraser , since you've been quite involved in the symbology import / export, your comments on this would be most helpful.

@wonder-sk , your comments would be great too since you've pretty much built this thing.

Others, please feel invited to go through this too.

NathanW2 commented:
Very impressive work.

On Sun, 13 Nov 2016 4:08 pm Mathieu Pellerin <notifications@github.com>
wrote:

> @anitagraser <https://github.com/anitagraser> , since you've been quite
> involved in the symbology import / export, your comments on this would be
> most helpful.
>
> @wonder-sk <https://github.com/wonder-sk> , your comments would be great
> too since you've pretty much built this thing.
>
> Others, please feel invited to go through this too.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/qgis/QGIS/pull/3755#issuecomment-260168981>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAXS3DolnC_ghifgQ9Xpt5T5zAXGTyuaks5q9qlhgaJpZM4Kwnjb>
> .
>

anitagraser commented:
Thanks for the notification! I'd love to hear from people who use the current grouping and tagging system. I have to admit that I never got into it.

I think it's a very good idea to support tagging directly when saving a new style. I also like the idea to have favorites. 

If only favorites are shown by default, how will we let users know that there are more symbols? Could we have a search in the  symbols list widget? And a system like in the Processing toolbox - where a message appears in the bottom if an algorithm is available in a deactivated provider?

Do you know if/how the changes affect the Resource Sharing plugin? cc @akbargumbira @elpaso

DelazJ commented:
Thanks @nirvn for this new feature. I suppose the advantage of tags vs groups is that a symbol can belong to multiple tags while it wasn't possible with groups. 

Some thoughts below:
Do you think a `not-tagged` category in the "Style Manager" could be helpful? maybe to help user manage their symbols.
Also, would the symbols list widget allow comma separated tags, instead of one tag in the combobox? sthg like `Basics, Infrastructure` to get listed all symbols from both tags.

I do not really get if it's only about symbols or also about styles (combination of symbolgy, label, diagram options...). I think 3.0 should be an opportunity to clean our vocabulary. For me, style is more than symbol (and a different thing), hence the "Add new style" dialog with tags is a bit confusing:
- shouldn't it precise that tags are related to symbol and not the style?
- and to which symbol (let's say in a categorized symbology) would the tags belong?

nirvn commented:
@anitagraser , thanks for your comments. Regarding the favorites shown by default, my wording might have been confusing. What I meant to say is that the "Favorites" category will be shown by default (i.e., this is what the symbols list widget will populate upon first opening). Users can click on the combo box at any time and select the "All Symbols" category. The combo box UX is also improved by the favorites, as it means QGIS doesn't ship with an empty combo box by default:
![untitled6](https://cloud.githubusercontent.com/assets/1728657/20251210/e18f7e98-aa49-11e6-9863-84a09d59700c.png)

When offering a populated combo box with "Favorites" as the selected item, we're increasing the chance of having to users go "hmm, what's in there other than favorites", which is when they hit the "All symbols" category and see all of the shipped symbols.

Assuming the resource sharing plugin uses the QgsStyle class, they might have to update the code a bit (I will add API breakage info when this PR is reviewed). 

Your mention of never having used tags/groups tells me I'm going in the right direction, thanks! :)

nirvn commented:
@DelazJ , yes, you are right, the main advantage with tags - which IMHO fits better in the context of saved symbols - is that you can attach many tag categories onto a symbol. The one thing we have to let go of when letting go of groups is the possibility of subgroups. That said, it can be emulated (and even behave better) by using tags with a "/" into it, i.e.:
* clients
* clients/company_a
* clients/company_b

This will result in the exact same listing int he symbols list widget, but with the advantage of being able to regroup all of the symbols tagged with "clients/company_a" and "clients/company_b" under the main "clients" tag. Just like Google does with gmail.

Regarding the combination of multiple tags, I'd suggest going through smartgroups ATM. Supporting multiple tags by typing a customized string in the combo box itself isn't something I'm planning to add.

Good point on the labeling of things, I'll try and improve the situation.

nyalldawson commented:
@nirvn If no symbols are favorited, what's shown by default? Does it fallback to showing all symbols or an empty list?

Great work on this PR!

nyalldawson commented on python/core/symbology-ng/qgsstyle.sip:
```diff
@@ -14,7 +14,7 @@ class QgsStyle : QObject
         database functions are being run.
         \sa group(), rename(), remove(), symbolsOfGroup(), symbolsWithTag(), symbolsOfSmartgroup()
      */
-    enum StyleEntity { SymbolEntity, GroupEntity, TagEntity, ColorrampEntity, SmartgroupEntity };
```
You'll need to update the api break docs to reflect these changes.

nyalldawson commented on python/core/symbology-ng/qgsstyle.sip:
```diff
@@ -110,6 +99,14 @@ class QgsStyle : QObject
      *  \return returns the success state of the operation
      */
     bool detagSymbol( StyleEntity type, QString symbol, QStringList tags );
+    //! clear the symbol from all attached tags
```
Some oddness here - looks like there's a duplicate doxygen tag, and the proper one should use `/**` not `/*!`

nyalldawson commented on src/core/symbology-ng/qgsstyle.h:
```diff
@@ -198,28 +193,17 @@ class CORE_EXPORT QgsStyle : public QObject
+     *  \return A QStringList of the symbol or colorramp names flagged as favorite
      */
-    QStringList symbolsOfGroup( StyleEntity type, int groupid );
+    QStringList symbolsOfFavorite( StyleEntity type );
```
Can this be made const?

nirvn commented:
@nyalldawson , 
* By default, QGIS will ship with {point,line,polygon} favorites, so users will not be presented with an empty list upon first use
* If users remove all favorites, then they will be served with an empty list

The alternative would be to serve all symbols when {favorites,tag,smartgroup} returns an empty list, but I'm not too hot to this idea.

nyalldawson commented on src/gui/symbology-ng/qgssymbolslistwidget.cpp:
```diff
@@ -69,24 +71,18 @@ QgsSymbolsListWidget::QgsSymbolsListWidget( QgsSymbol* symbol, QgsStyle* style,
   connect( viewSymbols->selectionModel(), SIGNAL( currentChanged( const QModelIndex &, const QModelIndex & ) ), this, SLOT( setSymbolFromStyle( const QModelIndex & ) ) );
 
   connect( mStyle, SIGNAL( symbolSaved( QString, QgsSymbol* ) ), this, SLOT( symbolAddedToStyle( QString, QgsSymbol* ) ) );
+  connect( mStyle, SIGNAL( groupsModified() ), this, SLOT( populateGroups() ) );
```
https://github.com/qgis/QGIS-Enhancement-Proposals/issues/77

nyalldawson reviewed (commented):
Code looks good to me... I'll leave the ux changes for Anita and others to discuss!

One thing - QgsStyle currently has quite good unit test coverage, can you please add tests for the new/changed behaviour introduced here?

nirvn commented:
@DelazJ , I've updated the code to insure that the save dialog better identifies whether it is saving a symbol (or a color ramp), instead of referring to the more generic "style". I've also updated the label of the save button embedded into the symbols list widget to explicitly refer to symbol:
![untitled](https://cloud.githubusercontent.com/assets/1728657/20252791/02fb0346-aa58-11e6-98a9-8c3e734f2af2.png)

NathanW2 commented:
Nice work.  The use of the wording Style there has always bugged me too.

Do you think it's much work that could also sort full styles in there?
Really just the output of a qml but just the style parts and not everything.

That would be very handy to have.

I'm happy to take a look at that once this is merged.

On Mon, Nov 14, 2016 at 1:49 PM, Mathieu Pellerin <notifications@github.com>
wrote:

> @DelazJ <https://github.com/DelazJ> , I've updated the code to insure
> that the save dialog better identifies whether it is saving a symbol (or a
> color ramp), instead of referring to the more generic "style". I've also
> updated the label of the save button embedded into the symbols list widget
> to explicitly refer to symbol:
> [image: untitled]
> <https://cloud.githubusercontent.com/assets/1728657/20252791/02fb0346-aa58-11e6-98a9-8c3e734f2af2.png>
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/qgis/QGIS/pull/3755#issuecomment-260242746>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAXS3B-4ilHEyWwf78EqY29SaHbCC1huks5q99pKgaJpZM4Kwnjb>
> .
>

nirvn commented:
@nyalldawson , your comments were addressed in this updated commit. Thanks for the review.

DelazJ commented:
> That said, it can be emulated (and even behave better) by using tags with a "/" into it

Sounds interesting to me. 
And thanks for fixing wording.

I had reported a couple of bug reports related to groups, when documenting this feature. Maybe your work make them irrelevant or fix them?
http://hub.qgis.org/issues/14074
http://hub.qgis.org/issues/14073
http://hub.qgis.org/issues/14072
http://hub.qgis.org/issues/14045 (rather feature request)
http://hub.qgis.org/issues/14047 (another feature request, more out of the scope)
Thanks again.

nirvn commented:
@DelazJ, this PR fixes issue 14074, 14073, and 14072. I'll close those when the PR is merged. Thanks for raising this.

nirvn commented:
Could we merge this today? It seems everybody agrees on the most "controversial" aspect of this PR (namely, getting rid of groups), and UX/code comments were addressed.

NathanW2 commented:
+1 to merge if people are happy.  I really like the new changes.

On Thu, Nov 17, 2016 at 11:07 AM, Mathieu Pellerin <notifications@github.com
> wrote:

> Could we merge this today? It seems everybody agrees on the most
> "controversial" aspect of this PR (namely, getting rid of groups), and
> UX/code comments were addressed.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/qgis/QGIS/pull/3755#issuecomment-261124073>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAXS3CkgrZVMXHKRo9AwjX8GROGq5V9qks5q-6jGgaJpZM4Kwnjb>
> .
>

nirvn commented:
@nyalldawson , thanks!

elpaso commented:
@nirvn @anitagraser @NathanW2  sorry for jumping in so late, I completely missed the question at https://github.com/qgis/QGIS/pull/3755#issuecomment-260176223 , unfortunately the QGIS Resource Sharing plugin relies completely upon groups, I discovered just today that I'm porting the plugin to QGIS3, that groups are gone.

My general concern here is that *tags* and *groups* are two completely different ways of organizing information, they are not mutually exclusive and they normal serve different purposes.

So, while I really like the introduction of *tags* I don't see a good reason for retiring the groups, unless you want to allow the creation of a tags taxonomy, which is kind of counter-intuitive but it would probably work from a information retrieval perspective.

For the QGIS Resource Sharing Plugin we can work around that, but I know for sure of one big public sector customers relying heavily on the tree-structure (taxonomy) of groups for keeping in order their (thousands of) symbology assets, they won't be happy about loosing that functionality.

Can we rethink this decision and maybe resurrect the groups?
